Add support for Windows broker (WAM)

mjcheetham opened the pull request:
Add support for broker-assisted authentication on Windows using "WAM" (Web Authentication Manager) as provided by the MSAL.Desktop library.

The `GCM_MSAUTH_USEBROKER` environment variable or the `credential.msauthUseBroker` configuration option will control if WAM is enabled or not. ~By default WAM _is_ enabled.~

This ended up being simpler then I expected. Looks like the VS AAD app registration has now been updated to have the required WAM redirect URI which was previously a technical blocker.

Fixes #211 

**Update:** WAM is now default _disabled_ due to concerns about the maturity of the UX and technology. cc: @bgavrilMS

dscho commented on docs/configuration.md:
```diff
@@ -258,6 +258,10 @@ Specify which authentication flow should be used when performing Microsoft authe
 
+**Note:** If [`credential.msauthUseBroker`](#credentialmsauthusebroker) is set
+to `true` and the operating system authentication broker is available, all flows
+will be delegated to the broker; this setting has no effect.
```
By "this setting" you mean `credential.msauthFlow`, and by "has no effect" you mean that `.msauthUseBroker` overrides `.msauthFlow`, right?

mjcheetham replied:
Yeah. Grammar here is hard. Any suggestions to make it clearer?

vtbassmatt replied:
Suggestion: truncate existing sentence at the `;`. Then add a second sentence:
> If both of those things are true, then the value of `credential.msauthFlow` has no effect.

dscho commented on docs/environment.md:
```diff
@@ -404,6 +404,10 @@ Specify which authentication flow should be used when performing Microsoft authe
 
+**Note:** If [`GCM_MSAUTH_USEBROKER`](#gcm_msauth_usebroker) is set to `true`
+and the operating system authentication broker is available, all flows will be
+delegated to the broker; this setting has no effect.
```
By "this setting" you mean `GCM_MSAUTH_FLOW`, correct? And by "has no effect" you mean to imply that `_USEBROKER` overrides `_FLOW`, correct?

vtbassmatt replied:
Same sentence suggestion as the other instance.

dscho commented on docs/environment.md:
````diff
@@ -427,6 +431,33 @@ export GCM_MSAUTH_FLOW="devicecode"
+export GCM_MSAUTH_USEBROKER="false"
+```
+
+**Also see: [credential.msauthUseBroker](configuration.md#credentialmsauthusebroker)**
````
Does the environment variable override the config setting, as is custom in Git?

mjcheetham replied:
Yes. Take a look at aka.ms/gcmcore-env:

![image](https://user-images.githubusercontent.com/5658207/114196614-7173aa00-9949-11eb-958d-c2d6f5450c18.png)

dscho commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -82,48 +97,61 @@ public MicrosoftAuthentication(ICommandContext context)
+
+                        default:
+                            goto case MicrosoftAuthenticationFlowType.Auto;
+                    }
```
The indentation change makes this harder to review than I'd like, but that cannot be helped, I guess.

mjcheetham replied:
Yeah. This is unfortunate 😢

derrickstolee replied:
![image](https://user-images.githubusercontent.com/570044/114253834-fd9dc580-9979-11eb-8132-faa0b197fc8b.png)

Reloading to ignore whitespace changes works great!

![image](https://user-images.githubusercontent.com/570044/114253836-02627980-997a-11eb-9805-624f5498e5eb.png)

dscho commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -205,6 +234,15 @@ private async Task<IPublicClientApplication> CreatePublicClientApplicationAsync(
+            if (enableBroker && PlatformUtils.IsWindows10())
+            {
+#if NETFRAMEWORK
+                appBuilder.WithExperimentalFeatures();
```
Experimental? That sounds safe... 
![image](https://user-images.githubusercontent.com/127790/113997468-7573cf00-9858-11eb-95f7-309f2b18c2df.png)

mjcheetham replied:
@bgavrilMS is this still needed for WAM support to be available?

bgavrilMS replied:
Experimental = public preview :). We're waiting to get some more traction on this feature before moving it to GA. I will discuss directly with Matthew if we make any breaking changes. IMO the only difficulties are around supporting Personal accounts.

dscho commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -360,6 +398,30 @@ public HttpClient GetHttpClient()
+            }
+#endif
+
+            return false;
```
Personal preference: if there are ways to return from a function early, do so, as it reduces cognitive load for the reader. In this instance:

```c#
if (!!context.SessionManager.IsDesktopSession || !PlatformUtils.IsWindows10())
    return false;

[...]
```

mjcheetham replied:
Updated.

dscho commented on src/shared/Microsoft.Git.CredentialManager/PlatformUtils.cs:
```diff
@@ -20,6 +20,26 @@ public static PlatformInformation GetPlatformInformation()
+            // https://github.com/dotnet/runtime/blob/6578f257e3be2e2144a65769706e981961f0130c/src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs#L110-L122
+            //
+            // Note that we cannot use Environment.OSVersion in .NET Framework (or Core versions less than 5.0) as
+            // The implementation in those versions "lies" about Windows versions > 8.1 if there is no application manifest.
```
s/The/the/

dscho reviewed (approved):
Do we need to be worried about [these warnings](https://github.com/microsoft/Git-Credential-Manager-Core/pull/323/commits/c81bd22191c2c31915cf7af4dbd25397d46c25f8#diff-d32249dde0548f66cf569037cf9a6fbe6d56f9637c5e454c09ba052294d41c85R339) (`StorageCreationPropertiesBuilder()` being obsolete)?

APACGAMONDE commented:
Congrats M. CHEETHAM!

vtbassmatt commented on docs/configuration.md:
```diff
@@ -275,6 +279,27 @@ git config --global credential.msauthFlow devicecode
+
+Use the operating system account manager where available.
+
+Defaults to the value `true`.
```
Nit suggestion:
> Defaults to `true`.

derrickstolee commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -28,7 +32,7 @@ public enum MicrosoftAuthenticationFlowType
         EmbeddedWebView = 1,
         SystemWebView = 2,
-        DeviceCode = 3
+        DeviceCode = 3,
```
Unrelated, but OK.

derrickstolee commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -9,6 +9,10 @@
 using Microsoft.Identity.Client;
 using Microsoft.Identity.Client.Extensions.Msal;
 
+#if NETFRAMEWORK
+using Microsoft.Identity.Client.Desktop;
+#endif
```
Is this going to be supported in .NET 5? Does adding this feature add more debt towards upgrading to .NET 5?

mjcheetham replied:
For a possible future move to .NET 5 on Windows, we'd be using the `net5.0-windows*` TFMs, of which the MSAL library includes the WAM pieces already.

The `MS.ID.Client.Desktop` package is there to bring those Windows-only dependencies in for .NET Framework (or .NET Core 3.x and below) without adding them to the core, x-plat `MS.ID.Client` library. Several of their customers had hard blockers on introducing things like a dependency on WPF or WinRT APIs in say, a use of MSAL in an ASP.NET application.

I've been round the houses on getting us on .NET 5 on Windows, but it's fraught with problems. Primarily around: a) trimming/size of the resulting application (bigger than the rest of G4W!), b) multi-targeting and down-level Windows support (win7+8.x), c) UI 'helper' assemblies.

cc: @bgavrilMS

bgavrilMS replied:
Yeah, ideally WAM support would be inside MSAL on all frameworks, however the depedencies on nuget packages that call into WinRT makes it impossible. In MSAL 5 we're thinking of making WAM the default option, and thus we'd need to separate MSAL into an MSAL for public client and MSAL for confidential client.

bgavrilMS commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -205,6 +234,15 @@ private async Task<IPublicClientApplication> CreatePublicClientApplicationAsync(
             }
 
+            // On Windows 10 & .NET Framework try and use the WAM broker
+            if (enableBroker && PlatformUtils.IsWindows10())
```
Broker is also available on Win Server 2016 and Win Server 2019, but the support there is a little iffy.

mjcheetham replied:
NT major version 10 should also capture Windows Server SKUs based on Windows 10.

bgavrilMS commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -205,6 +234,15 @@ private async Task<IPublicClientApplication> CreatePublicClientApplicationAsync(
+                appBuilder.WithWindowsBroker();
+#endif
+            }
+
```
for `StorageCreationPropertiesBuilder(cacheFileName, cacheDirectory, clientId)` just use `StorageCreationPropertiesBuilder(cacheFileName, cacheDirectory)`. The `clientId` param is only needed for situations where you want to monitor for accounts created / deleted via an event. This is used by VS to update their UI whenver an app (e.g. GCM) adds or removes an account. It's not being used by GCM.

mjcheetham replied:
Updated to remove this use of the deprecated ctor.

bgavrilMS commented on src/shared/Microsoft.Git.CredentialManager/PlatformUtils.cs:
```diff
@@ -20,6 +20,26 @@ public static PlatformInformation GetPlatformInformation()
+            }
+
+            // Implementation of version checking was taken from:
+            // https://github.com/dotnet/runtime/blob/6578f257e3be2e2144a65769706e981961f0130c/src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs#L110-L122
```
Very cool,. we use a nastier approach based on Registry for netframework.

mjcheetham replied:
Yeah I thought there has to be a better solution given the .NET folks invented all those `OperatingSystem.IsWindowsVersionOrGreater` etc APIs, and they correctly identify Windows 10 without manifests.

bgavrilMS commented on src/shared/Microsoft.Git.CredentialManager/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -360,6 +398,33 @@ public HttpClient GetHttpClient()
+        {
+#if NETFRAMEWORK
+            // We only support the broker on Windows 10 and require an interactive session
+            if (!Context.SessionManager.IsDesktopSession || !PlatformUtils.IsWindows10())
```
There is a helper for this in MSAL that we've added in collaboration with you: 

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/2cebdaec10f1991482ed7109a2fa34389a037299/src/client/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs#L318

mjcheetham replied:
Yeah, looking at this however we won't be able to use this sadly.. it's hanging off of the PCABuilder, and we need it before and after the builder exists 😢

mjcheetham commented:
> Do we need to be worried about [these warnings](https://github.com/microsoft/Git-Credential-Manager-Core/pull/323/commits/c81bd22191c2c31915cf7af4dbd25397d46c25f8#diff-d32249dde0548f66cf569037cf9a6fbe6d56f9637c5e454c09ba052294d41c85R339) (`StorageCreationPropertiesBuilder()` being obsolete)?

@dscho I've removed the warnings by dropping that extra (not needed) argument from the constructor in question.
We updated the dependency and in the new version, that API is marked obsolete.
